pkg/operator/ceph/object/bucket/provisioner.go

```diff
@@ -0,0 +1,413 @@
+		return err
+	}
+	if err = p.setObjectStorePort(sc); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (p *Provisioner) initializeDeleteOrRevoke(ob *bktv1alpha1.ObjectBucket) error {
+
+	sc, err := p.getStorageClassWithBackoff(ob.Spec.StorageClassName)
+	if err != nil {
+		return fmt.Errorf("failed to get storage class for OB %q: %v", ob.Name, err)
+	}
+
+	// set receiver fields from OB data
+	p.setBucketName(getBucketName(ob))
+	p.cephUserName = getCephUser(ob)
+	p.objectStoreName = getObjectStoreName(sc)
+	p.objectStoreNamespace = getObjectStoreNameSpace(sc)
+	err = p.setObjectContext()
+	if err != nil {
+		return err
+	}
+	if err = p.setObjectStoreDomainName(sc); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// Return the OB struct with minimal fields filled in.
+func (p *Provisioner) composeObjectBucket() *bktv1alpha1.ObjectBucket {
+
+	conn := &bktv1alpha1.Connection{
+		Endpoint: &bktv1alpha1.Endpoint{
+			BucketHost: p.storeDomainName,
+			BucketPort: int(p.storePort),
+			BucketName: p.bucketName,
+			Region:     p.region,
+			SSL:        false,
+		},
+		Authentication: &bktv1alpha1.Authentication{
+			AccessKeys: &bktv1alpha1.AccessKeys{
+				AccessKeyID:     p.accessKeyID,
+				SecretAccessKey: p.secretAccessKey,
+			},
+		},
+		AdditionalState: map[string]string{
+			cephUser: p.cephUserName,
+		},
+	}
+
+	return &bktv1alpha1.ObjectBucket{
+		Spec: bktv1alpha1.ObjectBucketSpec{
+			Connection: conn,
+		},
+	}
+}
+
+func (p *Provisioner) setObjectContext() error {
+	msg := "error building object.Context: store %s cannot be empty"
+	if p.objectStoreName == "" {
+		return fmt.Errorf(msg, "name")
+	} else if p.objectStoreNamespace == "" {
+		return fmt.Errorf(msg, "namespace")
+	}
+	p.objectContext = object.NewContext(p.context, p.objectStoreName, p.objectStoreNamespace)
+	return nil
+}
+
+// setObjectStoreDomainName sets the provisioner.storeDomainName and provisioner.port
+// must be called after setObjectStoreName and setObjectStoreNamespace
+func (p *Provisioner) setObjectStoreDomainName(sc *storagev1.StorageClass) error {
+
+	name := getObjectStoreName(sc)
+	namespace := getObjectStoreNameSpace(sc)
+	// make sure the object store actually exists
+	_, err := getObjectStore(p.context.RookClientset.CephV1(), p.objectStoreNamespace, p.objectStoreName)
+	if err != nil {
+		return err
+	}
+	p.storeDomainName = fmt.Sprintf("%s-%s.%s", prefixObjectStoreSvc, name, namespace)
+	return nil
+}
+
+func (p *Provisioner) setObjectStorePort(sc *storagev1.StorageClass) error {
+	name := getObjectStoreName(sc)
+	name = fmt.Sprintf("%s-%s", prefixObjectStoreSvc, name)
+	namespace := getObjectStoreNameSpace(sc)
+	// also ensure the service exists and get the appropriate clusterIP port
+	svc, err := getService(p.context.Clientset, namespace, name)
+	if err != nil {
+		return err
+	}
+	p.storePort = svc.Spec.Ports[0].Port
+	return nil
+}
+
+func (p *Provisioner) setObjectStoreName(sc *storagev1.StorageClass) {
+	const key = "objectStoreName"
```

Comment:
this const and the `"objectStoreNamespace"` one below are repeated in the util pkg also. is it possible to unify them?

Reply:
fixed